core/src/test/java/org/elasticsearch/cluster/ClusterHealthIT.java

```diff
@@ -62,10 +98,88 @@ public void testHealth() {
         assertThat(healthResponse.getIndices().get("test1").getStatus(), equalTo(ClusterHealthStatus.GREEN));
 
         logger.info("--> running cluster health on an index that does exists and an index that doesn't exists");
-        healthResponse = client().admin().cluster().prepareHealth("test1", "test2").setWaitForYellowStatus().setTimeout("1s").execute().actionGet();
+        healthResponse = client().admin().cluster().prepareHealth("test1", "test2")
+                                                   .setWaitForYellowStatus()
+                                                   .setTimeout("1s")
+                                                   .execute()
+                                                   .actionGet();
         assertThat(healthResponse.isTimedOut(), equalTo(true));
         assertThat(healthResponse.getStatus(), equalTo(ClusterHealthStatus.RED));
         assertThat(healthResponse.getIndices().get("test1").getStatus(), equalTo(ClusterHealthStatus.GREEN));
         assertThat(healthResponse.getIndices().size(), equalTo(1));
     }
-}
\ No newline at end of file
+
+    public void testHealthOnIndexCreation() throws Exception {
+        final int numNodes = randomIntBetween(2, 5);
+        logger.info("--> starting {} nodes", numNodes);
+        internalCluster().startNodesAsync(numNodes).get();
+
+        ClusterHealthResponse healthResponse = client().admin().cluster()
+                                                               .prepareHealth()
+                                                               .setWaitForGreenStatus()
+                                                               .setTimeout("10s")
+                                                               .execute()
+                                                               .actionGet();
+        assertThat(healthResponse.isTimedOut(), equalTo(false));
+        assertThat(healthResponse.getStatus(), equalTo(ClusterHealthStatus.GREEN));
+
+        // first, register a cluster state observer that checks cluster health
+        // upon index creation in the cluster state
+        final String masterNode = internalCluster().getMasterName();
+        final String indexName = "test-idx";
+        final ClusterService clusterService = internalCluster().clusterService(masterNode);
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, logger, threadPool.getThreadContext());
+        final ClusterStateObserver.ChangePredicate validationPredicate = new ClusterStateObserver.ValidationPredicate() {
+            @Override
+            protected boolean validate(ClusterState newState) {
+                return newState.status() == ClusterState.ClusterStateStatus.APPLIED
+                           && newState.metaData().hasIndex(indexName);
+            }
+        };
+
+        final ClusterStateObserver.Listener stateListener = new ClusterStateObserver.Listener() {
```

Comment:
We can have a simpler test case instead that does not rely on internals: have a thread create indices and another one running cluster health in parallel. See e.g. `IndexPrimaryRelocationIT` as inspiration. One thing to note is to increase `cluster.routing.allocation.node_initial_primaries_recoveries` so that all newly created indices are allocated right away.